Added a performance suite which stress tests the job API's

jleeh opened the pull request:
## Performance Tests
So, I've developed a test which bulk loads the following API's:

- Create Job
- View Job
- Create Job Run
- View Job Run

Levels of load being sent to each:

- 20/40/60 requests per second
- 4 seconds of load
- 80/160/240 requests in total for each

### Current Results
So from testing on my dev machine, with the stats of:

- Ryzen 5 1600 3.2Ghz
- 16GB DDR4

Results are as follows at 20 requests per second...

**Create Job:**
```
Average Latency: 53.658156ms
95th percentile: 60.7615ms
99th percentile: 61.1107ms
```

**View Job:**
```
Average Latency: 1.903137ms
95th percentile: 2.7378ms
99th percentile: 2.9195ms
```

**Create Job Runs:**
The time to complete runs statistic is the time elapsed for all the tasks to return either `completed` or `errored`.
```
Time to Complete Runs after API Calls: 21.2368091s (400ms on CI)
Average Latency: 1.658943ms
95th percentile: 2.6187ms
99th percentile: 2.8467ms
```

**View Job Runs:**
```
Average Latency: 13.320032ms
95th percentile: 15.1252ms
99th percentile: 16.2034ms
```

## Points for Discussion
- I've completely speculated on the requirements and I'm looking to tweak this with CI runs.
- Does 20/40/60 requests per second sound reasonable?
- Waiting upon an `errored` status within tasks. This is due to the following being thrown on the task run:
```json
{"level":"info","ts":1519484259.8783643,"caller":"services/job_runner.go:89","msg":"Finished current job run execution","job":"f55a8b8a4ae34eb9831eb064261dd380","run":"cb48eca67c384482948595dc16ee0df2","status":"errored","error":"EthMock: Method eth_getTransactionCount not registered"}
```
- Should this be enabled in CI? I'm presuming the Travis job may not have much resource to hand, so these times may increase significantly.

EDIT: Just seen they ran in CI and got a lot better statistics compared to my own machine, especially the waiting on tasks/jobs to complete. That took 400ms (wow) rather than 21s. Maybe that was related to the eth part failing for me?

EDIT 2: I've refactored the test so it data table drives the tests and bundles the utils with in.

dimroc commented:
Our logs track a very granular time stamp and could be repurposed to measure time spent in certain packages and methods ala [new relic](https://newrelic.com/). Example:

```
{"level":"debug","ts":1519687195.079926,"caller":"services/subscription.go:46","msg":"Listening for logs from address [all]"}
{"level":"debug","ts":1519687195.080074,"caller":"services/subscription.go:92","msg":"Received log from 0xABFf594e02F526FeFA31E87f0179776997490997 for job f6a8c1da33164564b34f5325f3520b93","log":{"address":"0xabff594e02f526fefa31e87f0179776997490997","topics":["0x06f4bf36b4e011a5c499cef1113c2d166800ce4013f6c2509cab1a0e92b83fb2","0x0000000000000000000000000000000000000000000000726571756573744944","0x6636613863316461333331363435363462333466353332356633353230623933"],"data":"0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000f7b2276616c7565223a22313030227d0000000000000000000000000000000000","blockNumber":"0x0","transactionHash":"0x0000000000000000000000000000000000000000000000000000000000000000","transactionIndex":"0x0","blockHash":"0x0000000000000000000000000000000000000000000000000000000000000000","logIndex":"0x0","removed":false},"job":{"id":"f6a8c1da33164564b34f5325f3520b93","initiators":[{"id":0,"jobId":"f6a8c1da33164564b34f5325f3520b93","type":"runlog","time":"0001-01-01T00:00:00Z","address":"0x0000000000000000000000000000000000000000"}],"tasks":[{}],"startAt":null,"endAt":null,"createdAt":"2018-02-26T18:19:55.077118-05:00"}}
{"level":"info","ts":1519687195.0809941,"caller":"services/job_runner.go:47","msg":"Starting job","job":"f6a8c1da33164564b34f5325f3520b93","run":"12bce71a0dc7416292d9d0757d68a037","status":"in progress"}
{"level":"debug","ts":1519687195.081135,"caller":"services/job_runner.go:64","msg":"Produced task run","tr":"TaskRun(92073c42e5cc45c0acc3474c542f6cd9,NoOp,completed,)"}
{"level":"info","ts":1519687195.082051,"caller":"services/job_runner.go:74","msg":"Task NoOp finished","task":0,"result":"","type":"NoOp","params":"{\"address\":\"0xABFf594e02F526FeFA31E87f0179776997490997\",\"dataPrefix\":\"0x0000000000000000000000000000000000000000000000726571756573744944\",\"functionSelector\":\"76005c26\",\"value\":\"100\"}","taskrun":"92073c42e5cc45c0acc3474c542f6cd9","status":""}
{"level":"info","ts":1519687195.0821362,"caller":"services/job_runner.go:89","msg":"Finished current job run execution","job":"f6a8c1da33164564b34f5325f3520b93","run":"12bce71a0dc7416292d9d0757d68a037","status":"completed"}
```
Maybe we could use new relic or the logs to profile our time spent into these stress tests to boost transparency.

dimroc commented on performance/api_load_test.go:
```diff
@@ -0,0 +1,212 @@
+
+func TestNodeUnderLoad(t *testing.T) {
+	Setup()
+	tests(t)
```
In general, we lean towards share nothing architectures. In common terms, that means that you don't share global `var`s and always pass and return copies.

The flow I see here is:
1. Define `tests` as type func(...)
2. Assign tests in `Setup`
3. Invoke `tests(t)`

Why not just have a method return `tests` and then invoke it, avoiding the global?

dimroc commented on performance/api_load_test.go:
```diff
@@ -0,0 +1,212 @@
+func CreateJobApiUnderLoad(
+	t *testing.T,
+) {
+	Setup()
```
It seems that `Setup` calls `CreateJobApiUnderLoad` and here we are with `CreateJobApiUnderLoad` calling `Setup`. You've run this to completion so there isn't a recursive loop but it does make us question the layout.

dimroc commented on performance/api_load_test.go:
```diff
@@ -0,0 +1,212 @@
+package performance
```
Our test packages are external to implementation, so `package performance_test`

jleeh commented on performance/api_load.go:
```diff
@@ -0,0 +1,144 @@
+		log.Fatal(err)
+	}
+	start := time.Now()
+	for {
```
Yeah, definitely a concern. I commented out the checking of this for anything above 20 requests per second because it was doing that. I couldn't find any other way to achieve the same result though. Is there any timestamps which provide a time a task/job was completed rather than when the run was created?

jleeh commented on performance/api_load_test.go:
```diff
@@ -0,0 +1,212 @@
+	tests = func(t *testing.T) {
+		for _, requirement := range requirements {
+			app, cleanup = cltest.NewApplicationWithKeyStore()
+			r = requirement
```
Only reason I did this was so the test methods had visibility to the requirement. If I added the requirement as a parameter, it broke being able to use that method within `t.Run()`.

Although, Go isn't my strong-suit and I'm learning it as I go along, please let me know if you know a better way to do it. Cause I'm not quite sure!

jleeh commented:
Thanks for the review @dimroc, and apologies for the standard of the code. Go isn't a language I've used an awful lot, more-so learning as I Go, ha.

I've actioned your comments, hopefully to a decent standard.

I agree with the high-level comment around whether it runs in the main-suite. The last thing I'd want to do is add a flaky test into your CI process. Maybe it could be part of an optional build process which just runs this test?

Ideally, we really need a fixed-spec environment with dedicated resources to manage this. Is that something you can define in Travis CI? I don't mind building a test environment for this, but I can't imagine how to cleanly implement that in an open-source environment.

Also, what's your views on the response time assertions? Do you feel the ms' I've set are too low or too little?

I'm not too much of a fan of using logged metrics for the assertion checking, as then it'd add complexity of having to calculate the metrics which are provided by default in vegata. Also, if for any reason that wasn't right with any future change, it would give us false positives rather than measuring e2e time in the library itself.

dimroc commented:
No worries, thanks for the ideas and the work. A lot of these questions I'll save for a discussion outside this PR with @se3000 and others.

> Also, what's your views on the response time assertions? Do you feel the ms' I've set are too low or too little?

I do think the ms response times are high, but it seemed that the CI box ran the perf tests faster than expected. I think we'll need to fix the polling loop before taking a hard look at the response times.

As for tracking the elapsed time of a job run, I have some ideas around this, both eliminating any sort of polling loop.

1. We add a final `HttpPost` task to a job that pushes to a HttpTest URL on completion.
2. Bake elapsed time as a field into JobRun, so viewing a job will give you the metric you're looking for.

The caveat with this setup is that we're using `vegeta` to attack the job controller, which makes sense, but it means we won't have access to `jobRun`. I had some more elaborate ideas around pushing to a channel on completion, but it wouldn't be accessible.

We'll chat about 2 but I think it'll make it into the backlog.

jleeh commented:
I've seen the commit around adding completion times to job runs @dimroc, so I'll re-factor this to use that and remove the polling loop.

Regarding getting access to the JobRuns, I can use `app.Store.JobRunsFor(job.ID)` like before and calculate a mean manually and percentile figures.

With the response times, they seem allover the place on Travis. When they first ran in CI, I was surprised at how quick it was compared to my own machine but then subsequent runs were magnitudes slower.

Let me know if you need any more changes after any discussions 😃

se3000 commented:
On the note of separating these out from the regular test suite: if you changed the tests to [benchmarks](https://golang.org/pkg/testing/#hdr-Benchmarks), which they pretty much are, they would only run when we flipped on the `-bench` flag.

jleeh commented:
Thanks for the feedback @dimroc and @se3000. I've updated the PR so it executes under `go bench` and the average time to complete a job run is calculated from subtracting the created from the completed time.

If there's any other tasks you want me to look at as part of this PR, let me know.

se3000 commented:
@jleeh [this](http://blog.campoy.cat/2014/03/github-and-go-forking-pull-requests-and.html) looks like a good way to avoid having to switch your import statements of a fork.

se3000 commented on internal/performance/api_load.go:
```diff
@@ -0,0 +1,162 @@
+			Method: "POST",
+			URL:    fmt.Sprintf("%s/v2/jobs", app.Server.URL),
+			Body:   schema,
+			Header: *GetBasicAuthHeader(app.Store),
```
In `GetBasicAuthHeader`, you're instantiating a struct and taking it's pointer, then returning a pointer, and taking its value here. Why not just return a value from `GetBasicAuthHeader`?

se3000 commented on internal/performance/api_load_test.go:
```diff
@@ -0,0 +1,201 @@
+func Setup() func(b *testing.B) {
+	 return func(b *testing.B) {
+		for _, req := range requirements {
+			app, cleanup = cltest.NewApplicationWithKeyStore()
```
Looks like `cleanup` doesn't need to be global, as it's only used in this function.

se3000 commented on internal/performance/api_load_test.go:
```diff
@@ -0,0 +1,201 @@
+	fmt.Printf("Average Latency: %s\n", metrics.Latencies.Mean)
+	fmt.Printf("95th percentile: %s\n", metrics.Latencies.P95)
+
+	assert.Condition(
```
Why use `assert.Condition`? Seems simpler to use `assert.True`.

jleeh commented:
I've refactored each of the tests into their own file. With not using globals, I couldn't share the app between the test. With the lines that added, I thought they'd be more readable and easier to build upon. This does mean the tests take longer as a whole, but at least gives the benefit of being able to run them independently.

se3000 commented on internal/performance/api_util_test.go:
```diff
@@ -0,0 +1,42 @@
+package performance
```
I'm not sure this needs to be tested. It's an the internal subpackage, where the code is all in service of benchmarking. The benchmarks should be sufficient, and if they don't work something fails. @dimroc might disagree with me on this though...?

dimroc replied:
I agree with you. I've moved on to the latest commit, just didn't want to leave you hanging.

se3000 commented on internal/performance/create_job_run_test.go:
```diff
@@ -0,0 +1,90 @@
+package performance
```
We write tests in an external package, like `performance_test`. This forces you to consider the API that other packages consume by not allowing you access to unexported functions and attributes. (Treat unexproted variables in a package like private variables in other languages.) I'm leaning towards treating benchmarks the same way.

se3000 commented on internal/performance/create_job_test.go:
```diff
@@ -0,0 +1,59 @@
+package performance
```
Splitting out the tests into separate functions makes a lot of sense to me. Splitting out the functions into separate files seems fine, but not necessarily idiomatic in the sense that you're no longer [grouping related code together](https://medium.com/@benbjohnson/structuring-applications-in-go-3b04be4ff091#982a). Not a big deal, just a suggestion.

se3000 commented on internal/performance/api_util_test.go:
```diff
@@ -0,0 +1,42 @@
+	defer cleanup()
+
+	targets := GetCreateJobTargets(app)
+	assert.Equal(t, targets[0].URL, fmt.Sprintf("%s/v2/jobs", app.Server.URL))
```
You'll probably need to rebase, and update the routes to `/specs` in place of `/jobs` everywhere. We received some feedback that it was unclear so we renamed it.

jleeh commented:
Changed it back to them all being in one file. I've added another unit test into the util for calculating the average duration and it's bumped up coverage nicely.

jleeh commented:
Just closing this PR based on what we discussed @se3000, with the complete refactor of this to be typical benchmark tests, didn't think it made sense to keep it under the same one.
